File: V2/com.questvisionstream/src/services/signaling/SignalingService.ts
import {
  BaseEventService,
  type ServiceActivationContext,
} from '@realitycollective/service-framework';
import { createLogger } from '../../util/logger';
import type {
  ISignalingService,
  SignalingEventMap,
  SignalingInbound,
  SignalingOutbound,
} from './ISignalingService';

const log = createLogger('Signaling');

export interface SignalingConfig {
  /** WebSocket URL, e.g. `ws://192.168.1.20:3000`. */
  readonly url: string;
  /** Auto-connect on service start (default true). */
  readonly autoConnect?: boolean;
  /** Reconnect backoff in ms; empty disables reconnect. */
  readonly reconnectBackoffMs?: readonly number[];
}

const DEFAULT_BACKOFF = [1000, 2000, 4000, 8000] as const;

/**
 * WebSocket close code the server sends when its connection cap evicts this
 * session in favour of a newer one (see `webrtc_server.py` `_enforce_connection_cap`).
 * We deliberately do NOT reconnect on this — see the onclose handler.
 */
const SUPERSEDED_CLOSE_CODE = 4000;

/**
 * WebSocket signaling transport. Speaks `webrtc_server.py`'s JSON protocol:
 * sends `offer`/`candidate`, receives `answer`/`candidate`. Reconnects with
 * bounded exponential backoff.
 */
export class SignalingService
  extends BaseEventService<SignalingEventMap, SignalingConfig>
  implements ISignalingService
{
  private socket: WebSocket | undefined;
  /** Open-promise for the in-flight CONNECTING socket, shared by all callers. */
  private pendingOpen: Promise<void> | undefined;
  private reconnectAttempt = 0;
  private reconnectTimer: ReturnType<typeof setTimeout> | undefined;
  private closedByUser = false;

  constructor(context: ServiceActivationContext<SignalingConfig>) {
    super(context);
  }

  private get cfg(): SignalingConfig {
    return this.serviceConfig;
  }

  get isConnected(): boolean {
    return this.socket?.readyState === WebSocket.OPEN;
  }

  override start(): void {
    // Attach a catch: an initial connect failure must not surface as an
    // unhandled rejection (onclose schedules the retry).
    if (this.cfg.autoConnect ?? true) {
      void this.connect().catch((err) => log.warn('Initial connect failed', err));
    }
  }

  /**
   * Contract: the returned promise resolves only once the socket is OPEN (so
   * `await connect()` always means "safe to send"), and rejects if this
   * connection attempt fails. Concurrent calls share the in-flight attempt.
   */
  connect(): Promise<void> {
    this.closedByUser = false;
    if (this.socket) {
      if (this.socket.readyState === WebSocket.OPEN) return Promise.resolve();
      if (this.socket.readyState === WebSocket.CONNECTING && this.pendingOpen) {
        return this.pendingOpen;
      }
    }

    const promise = new Promise<void>((resolve, reject) => {
      log.info(`Connecting to ${this.cfg.url}`);
      const socket = new WebSocket(this.cfg.url);
      this.socket = socket;
      let settled = false;

      socket.onopen = () => {
        if (this.socket !== socket) return; // superseded while connecting
        log.info('Connected');
        this.pendingOpen = undefined;
        this.reconnectAttempt = 0;
        this.emit('connected', undefined);
        settled = true;
        resolve();
      };
      socket.onmessage = (event) => {
        if (this.socket === socket) this.handleMessage(event.data);
      };
      socket.onerror = () => {
        if (this.socket === socket && socket.readyState !== WebSocket.OPEN) {
          this.pendingOpen = undefined;
          settled = true;
          reject(new Error('Signaling socket error'));
        }
      };
      socket.onclose = (event) => {
        // Settle the open-promise regardless of staleness so no caller hangs.
        if (!settled) {
          settled = true;
          if (this.pendingOpen === promise) this.pendingOpen = undefined;
          reject(new Error(`Signaling socket closed before opening (code ${event.code})`));
        }
        // A late close from a superseded socket must not look like a live
        // disconnect (and must not double-schedule reconnects).
        if (this.socket !== socket) return;
        this.socket = undefined;
        log.info(`Disconnected (code ${event.code})`);
        this.emit('disconnected', event.code);
        // 4000 = the server's connection cap superseded us with a NEWER
        // connection. Reconnecting would just supersede that one back — an
        // endless "who's connected" war between overlapping clients/tabs. Stay
        // down; the newest connection wins. (Reload to deliberately take over.)
        if (event.code === SUPERSEDED_CLOSE_CODE) {
          log.warn('Superseded by a newer connection — not reconnecting.');
          return;
        }
        if (!this.closedByUser) this.scheduleReconnect();
      };
    });
    this.pendingOpen = promise;
    return promise;
  }

  disconnect(): void {
    this.closedByUser = true;
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = undefined;
    }
    this.pendingOpen = undefined;
    this.socket?.close();
    this.socket = undefined;
  }

  send(message: SignalingOutbound): void {
    if (!this.isConnected) {
      log.warn('Dropping message; socket not open', message.type);
      return;
    }
    this.socket!.send(JSON.stringify(message));
  }

  override destroy(): void {
    this.disconnect();
    super.destroy();
  }

  private handleMessage(data: unknown): void {
    if (typeof data !== 'string') return;
    let parsed: SignalingInbound;
    try {
      parsed = JSON.parse(data) as SignalingInbound;
    } catch {
      log.warn('Ignoring non-JSON signaling message');
      return;
    }
    if (parsed?.type === 'answer' || parsed?.type === 'candidate') {
      this.emit('message', parsed);
    }
  }

  private scheduleReconnect(): void {
    const backoff = this.cfg.reconnectBackoffMs ?? DEFAULT_BACKOFF;
    if (backoff.length === 0) return;
    const delay = backoff[Math.min(this.reconnectAttempt, backoff.length - 1)]!;
    this.reconnectAttempt += 1;
    log.info(`Reconnecting in ${delay}ms (attempt ${this.reconnectAttempt})`);
    this.reconnectTimer = setTimeout(() => {
      void this.connect().catch(() => {
        /* onclose reschedules */
      });
    }, delay);
  }
}

File: V2/com.questvisionstream.unity/Runtime/Services/Signaling/SignalingService.cs
// Copyright (c) Simon Jackson (SimonDarksideJ). All rights reserved.
// Licensed under the MIT License. See LICENSE in the repository root for license information.

using System;
using QuestVisionStream.Protocol;
using RealityCollective.ServiceFramework.Services;
using UnityEngine;
using Utilities.WebSockets;

namespace QuestVisionStream.Services
{
    /// <summary>
    /// <see cref="ISignalingService"/> over <c>com.utilities.websockets</c> (all
    /// events arrive on the Unity main thread, so no locking is needed here).
    /// Ported from the hardened V2 TypeScript <c>SignalingService</c>: bounded
    /// backoff, one reconnect in flight at a time, stale-socket guards, no
    /// reconnect on the server's deliberate-eviction close codes (4000/4001).
    /// </summary>
    [System.Runtime.InteropServices.Guid("9ef447c2-21f5-4b5b-acbb-1510b4d6bc04")]
    public class SignalingService : BaseServiceWithConstructor, ISignalingService
    {
        private readonly SignalingServiceProfile profile;

        /// <summary>
        /// Per-app-launch connection identity. A reconnect carrying the same cid
        /// cleanly replaces our own half-open prior session (server close 4001)
        /// instead of fighting the connection cap.
        /// </summary>
        private readonly string connectionId = Guid.NewGuid().ToString("N");

        private WebSocket socket;
        private bool closedByUser;
        private int reconnectAttempt;
        private float reconnectAtRealtime = -1f;
        private float nextKeepAliveRealtime = -1f;
        private SignalingCloseInfo? lastCloseInfo;

        public SignalingService(string name, uint priority, SignalingServiceProfile profile)
            : base(name, priority)
        {
            this.profile = profile != null ? profile : throw new ArgumentNullException(nameof(profile));
        }

        public event Action Connected;
        public event Action<int> Disconnected;
        public event Action<string> AnswerReceived;
        public event Action<IceCandidateMessage> CandidateReceived;

        public bool IsConnected => socket != null && socket.State == State.Open;

        public SignalingCloseInfo? LastCloseInfo => lastCloseInfo;

        /// <inheritdoc />
        public override void Start()
        {
            base.Start();

            if (profile.AutoConnect)
            {
                Connect();
            }
        }

        /// <inheritdoc />
        public override void Update()
        {
            base.Update();

            var now = Time.realtimeSinceStartup;

            // Scheduled reconnect (single timer — never stacked).
            if (reconnectAtRealtime >= 0 && now >= reconnectAtRealtime)
            {
                reconnectAtRealtime = -1f;
                Connect();
            }

            // Application-level keepalive: __keepalive generates traffic without server log spam.
            if (IsConnected && profile.KeepAliveIntervalSeconds > 0 && now >= nextKeepAliveRealtime)
            {
                nextKeepAliveRealtime = now + profile.KeepAliveIntervalSeconds;
                SendStatus("__keepalive", string.Empty);
            }
        }

        /// <inheritdoc />
        public void Connect()
        {
            closedByUser = false;
            reconnectAtRealtime = -1f;

            if (socket != null &&
                (socket.State == State.Open || socket.State == State.Connecting))
            {
                // Never open a second socket while one is live or still connecting —
                // overlapping sockets are exactly how ghost sessions were created.
                return;
            }

            DisposeSocket();

            var url = BuildUrl();
            Debug.Log($"[QVS:Signaling] Connecting to {profile.ServerUrl}");
            var newSocket = new WebSocket(url);
            socket = newSocket;

            newSocket.OnOpen += () =>
            {
                if (socket != newSocket) { return; } // superseded while connecting
                Debug.Log("[QVS:Signaling] Connected");
                reconnectAttempt = 0;
                nextKeepAliveRealtime = Time.realtimeSinceStartup + profile.KeepAliveIntervalSeconds;
                Connected?.Invoke();
            };

            newSocket.OnMessage += dataFrame =>
            {
                if (socket != newSocket || dataFrame.Type != OpCode.Text) { return; }
                HandleMessage(dataFrame.Text);
            };

            newSocket.OnError += exception =>
            {
                if (socket != newSocket) { return; }
                Debug.LogWarning($"[QVS:Signaling] Socket error: {exception.Message}");
                // OnClose follows and drives the retry — nothing else to do here.
            };

            newSocket.OnClose += (code, reason) =>
            {
                // A late close from a superseded socket must not look like a live
                // disconnect (or double-schedule reconnects).
                if (socket != newSocket) { return; }
                socket = null;
                newSocket.Dispose();

                var closeCode = (int)code;
                lastCloseInfo = new SignalingCloseInfo(closeCode, reason);
                Debug.LogWarning($"[QVS:Signaling] Disconnected: {lastCloseInfo}");
                Disconnected?.Invoke(closeCode);

                // 4000 = superseded by a different client; 4001 = replaced by our own
                // reconnect. The server deliberately closed us in favour of another
                // connection — reconnecting would just fight it. Stay down.
                if (ServerCloseCodes.SuppressesReconnect(closeCode))
                {
                    Debug.LogWarning($"[QVS:Signaling] Closed by server ({ServerCloseCodes.Describe(closeCode)}) — not reconnecting.");
                    return;
                }

                if (!closedByUser)
                {
                    ScheduleReconnect();
                }
            };

            newSocket.Connect();
        }

        /// <inheritdoc />
        public void Disconnect()
        {
            closedByUser = true;
            reconnectAtRealtime = -1f;

            if (socket != null)
            {
                var closing = socket;
                socket = null;
                closing.Close();
                closing.Dispose();
            }
        }

        /// <inheritdoc />
        public void SendOffer(string sdp) => Send(SignalingMessages.Offer(sdp));

        /// <inheritdoc />
        public void SendCandidate(string candidate, string sdpMid, int sdpMLineIndex)
            => Send(SignalingMessages.Candidate(candidate, sdpMid, sdpMLineIndex));

        /// <inheritdoc />
        public void SendStatus(string field, string value)
            => Send(SignalingMessages.Status(field, value), quiet: true);

        /// <inheritdoc />
        public override void Destroy()
        {
            Disconnect();
            base.Destroy();
        }

        private void Send(string message, bool quiet = false)
        {
            if (!IsConnected)
            {
                if (!quiet)
                {
                    Debug.LogWarning("[QVS:Signaling] Dropping message; socket not open");
                }
                return;
            }

            _ = socket.SendAsync(message);
        }

        private void HandleMessage(string json)
        {
            // Unknown/malformed messages are skipped, never fatal (additive wire rule).
            switch (SignalingMessages.Parse(json, out var sdp, out var candidate))
            {
                case SignalingMessageKind.Answer:
                    AnswerReceived?.Invoke(sdp);
                    break;
                case SignalingMessageKind.Candidate:
                    CandidateReceived?.Invoke(candidate);
                    break;
            }
        }

        private void ScheduleReconnect()
        {
            if (reconnectAtRealtime >= 0)
            {
                return; // one reconnect in flight at a time
            }

            var backoff = profile.ReconnectBackoffSeconds;
            if (backoff == null || backoff.Length == 0)
            {
                return;
            }

            var delay = backoff[Mathf.Min(reconnectAttempt, backoff.Length - 1)];
            reconnectAttempt++;
            Debug.Log($"[QVS:Signaling] Reconnecting in {delay:0.#}s (attempt {reconnectAttempt})");
            reconnectAtRealtime = Time.realtimeSinceStartup + delay;
        }

        private Uri BuildUrl()
        {
            var url = profile.ServerUrl.TrimEnd('/');
            var separator = url.Contains("?") ? "&" : "?";
            url = $"{url}{separator}cid={connectionId}";
            if (!string.IsNullOrEmpty(profile.AuthToken))
            {
                url += $"&token={Uri.EscapeDataString(profile.AuthToken)}";
            }

            return new Uri(url);
        }

        private void DisposeSocket()
        {
            if (socket == null)
            {
                return;
            }

            var old = socket;
            socket = null;
            try
            {
                old.Dispose();
            }
            catch
            {
                // Disposing a partially-opened socket must never take the service down.
            }
        }
    }
}
